File: internal/ui/summary.go
package ui

import (
	"fmt"
	"strings"
	"time"

	"github.com/charmbracelet/lipgloss"

	"github.com/nlebedevinc/pulse/internal/probe"
	"github.com/nlebedevinc/pulse/internal/stats"
)

// Summary renders the end-of-session report printed after the TUI exits.
func Summary(checks *probe.Checks, t *stats.Tracker, kind string, elapsed time.Duration) string {
	var b strings.Builder

	target := checks.Target
	if checks.IP != nil && checks.IP.IP.String() != target {
		target += dim.Render(" (" + checks.IP.IP.String() + ")")
	}
	b.WriteString(brand.Render("pulse") + " summary — " + target + "\n")

	if t.Sent() == 0 {
		b.WriteString(dim.Render("  no probes sent") + "\n")
		return b.String()
	}

	b.WriteString(fmt.Sprintf("  %d probes over %s via %s · %d lost (%s)\n",
		t.Sent(), elapsed, kind, t.Lost(),
		lossStyle(t.Loss()).Render(fmtLoss(t.Loss()))))

	if t.Recv() > 0 {
		b.WriteString("  rtt min/avg/max " +
			value.Render(fmt.Sprintf("%s/%s/%s", fmtMs(t.Min()), fmtMs(t.Avg()), fmtMs(t.Max()))) +
			dim.Render(" · ") + dim.Render("p95 ") + value.Render(fmtMs(t.Percentile(95))) +
			dim.Render(" · ") + dim.Render("jitter ") + value.Render(fmtMs(t.Jitter())) + "\n")
	}

	grade, reason := t.Verdict()
	b.WriteString("  verdict: " + gradeStyle(grade).Render(grade.String()) +
		dim.Render(" — "+reason) + "\n")
	return b.String()
}

func gradeStyle(g stats.Grade) lipgloss.Style {
	switch g {
	case stats.Excellent:
		return ok.Bold(true)
	case stats.Good:
		return ok
	case stats.Degraded:
		return warn
	default:
		return bad.Bold(true)
	}
}

File: src/ui/summary.rs
//! The end-of-session report printed after the TUI exits.

use std::time::Duration;

use crate::checks::Checks;
use crate::stats::Tracker;
use crate::ui::format::{fmt_loss, fmt_ms, fmt_secs, loss_style};
use crate::ui::styles::{Style, BAD_BOLD, BRAND, DIM, OK, OK_BOLD, VALUE, WARN};
use crate::verdict::Grade;

pub fn summary(checks: &Checks, t: &Tracker, kind: &str, elapsed: Duration) -> String {
    let mut b = String::new();

    let mut target = checks.target.clone();
    if let Some(ip) = checks.ip {
        let ip = ip.to_string();
        if ip != target {
            target.push_str(&DIM.render(&format!(" ({ip})")));
        }
    }
    b.push_str(&format!("{} summary — {target}\n", BRAND.render("pulse")));

    if t.sent() == 0 {
        b.push_str(&DIM.render("  no probes sent"));
        b.push('\n');
        return b;
    }

    b.push_str(&format!(
        "  {} probes over {} via {} · {} lost ({})\n",
        t.sent(),
        fmt_secs(elapsed),
        kind,
        t.lost(),
        loss_style(t.loss()).render(&fmt_loss(t.loss()))
    ));

    if t.recv() > 0 {
        b.push_str("  rtt min/avg/max ");
        b.push_str(&VALUE.render(&format!(
            "{}/{}/{}",
            fmt_ms(t.min()),
            fmt_ms(t.avg()),
            fmt_ms(t.max())
        )));
        b.push_str(&DIM.render(" · "));
        b.push_str(&DIM.render("p95 "));
        b.push_str(&VALUE.render(&fmt_ms(t.percentile(95.0))));
        b.push_str(&DIM.render(" · "));
        b.push_str(&DIM.render("jitter "));
        b.push_str(&VALUE.render(&fmt_ms(t.jitter())));
        b.push('\n');
    }

    let (grade, reason) = t.verdict();
    b.push_str("  verdict: ");
    b.push_str(&grade_style(grade).render(&grade.to_string()));
    b.push_str(&DIM.render(&format!(" — {reason}")));
    b.push('\n');
    b
}

fn grade_style(g: Grade) -> Style {
    match g {
        Grade::Excellent => OK_BOLD,
        Grade::Good => OK,
        Grade::Degraded => WARN,
        Grade::Poor => BAD_BOLD,
    }
}
